sklearn/neural_network/_stochastic_optimizers.py

```diff
@@ -0,0 +1,259 @@
+        msg : str
+            Message passed in for verbose output
+        verbose : bool
+            Print message to stdin if True
+
+        Returns
+        -------
+        is_stopping : bool
+            True if traning needs to stop
+        """
+        if verbose:
+            print(msg + " Stopping.")
+        return True
+
+    def _get_updates(self, grads):
+        """Get the amount to update to params with given gradients
+
+        Parameters
+        ----------
+        grads : list, length = len(coefs_) + len(intercepts_)
+            Containing gradients with respect to coefs_ and intercepts_ in MLP
+            model. So length should be aligned with params
+
+        Returns
+        -------
+        updates : list, length = len(grads)
+            The amount to add to params
+        """
+
+        return [-self.learning_rate * grad for grad in grads]
+
+
+class SGDOptimizer(BaseOptimizer):
+    """Stochastic gradient descent optimizer with momentum
+
+    Parameters
+    ----------
+    params : list, length = len(coefs_) + len(intercepts_)
+        The list containing coefs_ and intercepts_ concatenated in MLP model.
+        Used for initializing velocities and updating params
+
+    learning_rate_init : float, optional, default 0.1
+        The initial learning rate used. It controls the step-size in updating
+        the weights
+
+    lr_schedule : {'constant', 'adaptive', 'invscaling'}, default 'constant'
+        Learning rate schedule for weight updates.
+
+        -'constant', is a constant learning rate given by
+         'learning_rate_init'.
+
+        -'invscaling' gradually decreases the learning rate 'learning_rate_' at
+          each time step 't' using an inverse scaling exponent of 'power_t'.
+          learning_rate_ = learning_rate_init / pow(t, power_t)
+
+        -'adaptive', keeps the learning rate constant to
+         'learning_rate_init' as long as the training keeps decreasing.
+         Each time 2 consecitive epochs fail to decrease the training loss by
+         tol, or fail to increase validation score by tol if 'early_stopping'
+         is on, the current learning rate is divided by 5.
+
+    momentum : float, optional, default 0.9
+        Value of momentum used, must be larger than or equal to 0
+
+    nesterov : bool, optional, default True
+        Whether to use nesterov's momentum or not. Use nesterov's if True
+
+    Attributes
+    ----------
+    learning_rate : float
+        the current learning rate
+
+    velocities : list, length = len(params)
+        velocities that are used to update params
+    """
+
+    def __init__(self, params, learning_rate_init=0.1, lr_schedule='constant',
+                 momentum=0.9, nesterov=True, power_t=0.5, **kwargs):
+        super(SGDOptimizer, self).__init__(params, learning_rate_init)
+
+        self.lr_schedule = lr_schedule
+        self.momentum = momentum
+        self.nesterov = nesterov
+        self.power_t = power_t
```

Comment:
BaseOptimizer and SGDOptimizer seem basically the same - BaseOptimizer could definitely get folded into SGDOptimizer I am pretty sure